### ISSUE TITLE
Support Node.js v4 plan

### ISSUE THREAD

**Issue by okkez** [fluent/fluent-logger-node#38]
I'm investigating that we should support which versions of Node.js.

This library depends on [msgpack-node](https://github.com/msgpack/msgpack-node), so we can support same versions that is supported by msgpack-node.

msgpack-node supports Node.js v0.12 or later. I've tested to build msgpack-node(1.0.2 and HEAD) following versions:

Version | 1.0.2 | HEAD
-----------|--------|----------
v0.10.40 | NG | NG
v0.12.7 | OK | OK
iojs-v1.8.4 | OK | OK
iojs-v2.5.0 | OK | OK
iojs-v3.3.1 | OK | OK
v4.0.0 | OK | OK
v4.1.2 | OK | OK
v4.2.1 | OK | OK

msgpack-node v0.2.7 supports Node.js v0.10.x.

[LTS](https://github.com/nodejs/LTS) says that current LTS versions are v0.12.x and v4.2.x. v0.10.x has been maintenance mode since 2015-10-01.

Therefore I think that we can support Node.js versions as following:

Branch name | Supported version | Comment
------------------|-------------------------|-------------
master | v0.12 or later | Support actively. Using latest version of msgpack-node.
v0.10  | v0.10.x | Bug fix only. Using old version of msgpack-node.
nodejs-4 | Drop | This branch does not support msgpack-node.

How about this plan? @edsiper @repeatedly

**Comment by okkez** [comment]
Additional information.

msgpack5 is too slow, so I want to revert msgpack5 related changes.

See http://yukidarake.hateblo.jp/entry/2015/08/10/194319 (Written in Japanese)

**Comment by edsiper** [comment]
thanks for raising this topic. 

How complex is to port to the new library ?

**Comment by okkez** [comment]
It is very easy to support msgpack-1.0.2.
Just revert msgpack5 related changes in 6f2a3f6 and adjust package.json.

I have patches in my laptop and I will work for this in next week.

**Comment by edsiper** [comment]
that would be great. I'd definitely support this change.

**Comment by repeatedly** [comment]
In fluent-logger-node, we use only encoding.
Could you check only `pack`/`encode` performance?
If msgpack5 is too slow, vote for changing to msgpack.

**Comment by okkez** [comment]
Check performance.
https://gist.github.com/okkez/2b77b32ad3a966446347

msgpack is faster than msgpack5.

```text
$ node benchmark.js
msgpack5: 856ms
msgpack: 29ms
```

**Comment by repeatedly** [comment]
I see. One concern is msgpack5 has extension type support.
Does msgpack have this feature?

**Comment by okkez** [comment]
FYI: [kawanet/msgpack-lite](https://github.com/kawanet/msgpack-lite "kawanet/msgpack-lite")

msgpack-lite is faster than msgpack.
msgpack-lite supports extension type.

**Comment by okkez** [comment]
Update benchmark script and result.
https://gist.github.com/okkez/2b77b32ad3a966446347

```text
$ node ./benchmark.js
msgpack5: 821ms
msgpack: 36ms
msgpack-lite: 23ms
```

**Comment by okkez** [comment]
> Does msgpack have this feature?

Maybe msgpack does not support extension type feature.
I've checked JavaScript APIs and README.md.

**Comment by okkez** [comment]
I've created branch https://github.com/okkez/fluent-logger-node/tree/use-msgpack-lite
It is very easy to change back end library to process msgpack.
Pass all tests with msgpack-lite for now.

How about msgpack-lite for fluent-logger-node?
If we use msgpack-lite, we can support all node versions in master branch for now.

**Comment by repeatedly** [comment]
:+1:  > msgpack-lite

**Comment by okkez** [comment]
I've released new version 0.4.1 using msgpack-lite.

**Comment by kawanet** [comment]
:smiley:
